### ISSUE TITLE
Setting FlushMode to Immediate will hit assert 0x16a

### ISSUE THREAD

**Issue by agarwal-navin** [microsoft/FluidFramework#9398]
When TurnBased is the default FlushMode, setting it manually to Immediate will hit assert 0x16a here - https://github.com/microsoft/FluidFramework/blob/71cab8a8b36e91d97dcbd0d2c5ed16c17ddf5e91/packages/runtime/container-runtime/src/pendingStateManager.ts#L376.

The bug is that the Immediate FlushMode is added to the pending state queue. When a local message is ack'd after this, it hits this assert which expects the FlushMode to be TurnBased. From a little debugging, it seems like the pending state manager's validation were based around the assumption that Immediate is the default flush mode but when this changes, the validation may not be correct.

**Comment by vladsud** [comment]
Assigning to current month as that's regression. We should assess if hot fix is required.

**Comment by andre4i** [comment]
Do we know the original intention of that assertion? What is it actually validating when it's checking for the value of the pending flushmode state? @agarwal-navin

**Comment by agarwal-navin** [comment]
Added a test for that repros this bug - https://github.com/microsoft/FluidFramework/pull/9402.

**Comment by agarwal-navin** [comment]
> Do we know the original intention of that assertion? What is it actually validating when it's checking for the value of the pending flushmode state? @agarwal-navin

Looking at the history to determine this.

**Comment by agarwal-navin** [comment]
The pending state manager has two responsibilities:
1. Replay pending message in order and preserve batch state of these messages during replay.
2. Validate that the batch information is preserved when ops round trip from the server.

When `Immediate` is the default `FlushMode`, this is what the tracking looks like:
- `FlushMode.Immediate` - This is tracked only if there is a message before it. Since we start with `Immediate`, we would need to replay and set it again only if the `FlushMode` was set to `TurnBased`, some ops were sent and it was set back to `Immediate`. Note that we do not track the default `FlushMode`.
- `FlushMode.TurnBased` - This is always tracked. But if the `FlushMode` is changed again without sending any ops, we remove this from the pending state queue since there is no point in tracking it.
- Basically, the `FlushMode` is tracked only if when its value changes and there are ops in between.

Based on this, the validation works as follows:
- When it first sees a `FlushMode`, it validates its `TurnBased` since we started with `Immediate`. This also indicates the beginning of a new batch. This `FlushMode` is remove from the queue.
- When a batch ends, it's either because `flush` was called or because `FlushMode` was set to `Immediate`. It its the latter, it removes `FlushMode.Immediate` from the queue.
- The next `FlushMode` it sees should be `TurnBased` so it validates that again and so on.

The bug here is that now the first `FlushMode` it sees can be `Immediate` and then the validation fails. After that, the same logic applies where the `FlushMode` will alternate.

**Comment by agarwal-navin** [comment]
However, I do think that the validation is complex and can probably be simplified. We can track all the `FlushMode` states irrespective of whether there were ops in between.

**Comment by agarwal-navin** [comment]
I am also wondering if the processing of remote op can have an impact since it empties the queue - https://github.com/microsoft/FluidFramework/blob/8e864f77c902c15472ffe1d4cd59362c106477ca/packages/runtime/container-runtime/src/pendingStateManager.ts#L285.

My guess is that it only empties the queue if all it contains were stashed ops but it's not clear if that's what is happening. @wes-carlson @vladsud any idea?

**Comment by andre4i** [comment]
I have some trouble understanding on the conceptual level whether this particular assert is out of place with the current invariants (default flushmode) or if it still guards the system from ending up in an inconsistent state. I can't find any instances of it in Kusto (although that doesn't tell us much). If we have good proof that the latter is not the case, we can just remove it and backlog an item to simplify this validation.

**Comment by agarwal-navin** [comment]
My only concern with removing it is whether it may end up hitting other asserts in that file. I think it should be fine but we should do some testing to validate that. For instance, starting with either of the modes and then switching between them a few times in the same document and making sure that things are working as expected.

**Comment by wes-carlson** [comment]
>My guess is that it only empties the queue if all it contains were stashed ops but it's not clear if that's what is happening.

This is only reached when the remote op is from a previous container that we got stashed ops from. It will stop and return the first op it sees so it shouldn't empty the queue.
